src/lib.rs

```diff
@@ -0,0 +1,489 @@
+lazy_static! {
+    static ref HOT_COUNTS: RwLock<HashMap<JitLocation, HotCount>> = RwLock::new(HashMap::new());
+}
+
+/// Get the hot count for the given location.
+fn hot_count(loc: JitLocation) -> HotCount {
+    let ret = match HOT_COUNTS.read().unwrap().get(&loc) {
+        Some(v) => *v,
+        None => 0,
+    };
+    ret
+}
+
+/// Atomically increment and return the hot count for the given location.
+fn incr_hot_count(loc: JitLocation) -> HotCount {
+    // XXX don't store locations which cannot be loop heads.
+    let mut counts = HOT_COUNTS.write().unwrap();
+    let loc_count = counts.entry(loc).or_insert(0);
+    *loc_count += 1;
+    *loc_count
+}
+
+/// Thread-specific meta-tracer context.
+///
+/// Each thread has one of instance of this inside a thread-local.
+#[derive(Debug)]
+pub struct JitThreadContext {
+    phase: JitPhase,
+    // XXX PT packet buffer goes here.
+}
+
+impl JitThreadContext {
+    pub fn new() -> Self {
+        Self {
+            phase: JitPhase::Interpreting,
+        }
+    }
+
+    /// Resets the per-thread JIT state to defaults.
+    #[cfg(test)]
+    fn reset(&mut self) {
+        self.phase = JitPhase::Interpreting;
+    }
+
+    /// Get the current thread's JIT phase.
+    fn current_phase() -> JitPhase {
+        let phase = JIT_THREAD_CX.with(|k| {
+            k.borrow().phase.clone()
+        });
+        phase
+    }
+
+    /// Transition to a new JIT phase.
+    fn phase_transition(phase: JitPhase) {
+        JIT_THREAD_CX.with(|k| {
+            k.borrow_mut().phase = phase;
+        });
+    }
+
+    /// Compile a trace from the raw PT packets (in parallel with the interpreter)..
+    ///
+    // XXX eventually we probably want a thread pool, as having an unbounded number of threads
+    // is probably bad.
+    fn compile_trace(loc: JitLocation) {
+        let name = format!("compile_loc_{}", loc);
+        debug!("new compilation thread: {}", name);
+        // By discarding the resultant JoinHandle here, we are creating "detached threads".
+        named_thread(name, move || {
+                // XXX This is where we would invoke the real trace compiler. For now we just
+                // pretend to, and insert a dummy trace.
+                set_trace(loc, JitTrace::FinalisedTrace);
+                debug!("compilation thread {} terminating", debug_thread_name());
+        });
+    }
+}
+
+thread_local! {
+    static JIT_THREAD_CX: RefCell<JitThreadContext> = RefCell::new(JitThreadContext::new());
+}
+
+/// Indicates what a thread is doing at any given time.
+#[derive(Debug, Eq, PartialEq, Clone)]
+enum JitPhase {
+    Interpreting,               // Interpreting and profiling.
+    Tracing(JitLocation),       // Recording a trace from the specified location.
+    RunningTrace(JitLocation),  // Running a trace starting from the specified location.
+}
+
+/// A machine code trace for a location.
+#[derive(Debug, PartialEq, Eq)]
+enum JitTrace {
```

Comment:
I think these entries should just be `Pending` and `Ready` (brevity is a virtue and all that).